File: app/src/main/java/apps/sai/com/imageresizer/myimages/MyImagesAdaptor.java
package apps.sai.com.imageresizer.myimages;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import apps.sai.com.imageresizer.BaseFragment;
import apps.sai.com.imageresizer.R;
import apps.sai.com.imageresizer.data.DataApi;
import apps.sai.com.imageresizer.data.FileApi;
import apps.sai.com.imageresizer.data.ImageInfo;
import apps.sai.com.imageresizer.listener.MultipleImageProcessingDialog;
import apps.sai.com.imageresizer.resize.ResizeFragment;
import apps.sai.com.imageresizer.settings.SettingsManager;
import apps.sai.com.imageresizer.util.ImageDetailFragment;
import apps.sai.com.imageresizer.util.ImageInfoLoadingTask;
import apps.sai.com.imageresizer.util.SpacesItemDecoration;
import apps.sai.com.imageresizer.util.Utils;

/**
 * Created by sailesh on 20/01/18.
 */

public class MyImagesAdaptor extends RecyclerView.Adapter<RecyclerView.ViewHolder> implements ResizeFragment.OnImagedSavedListener {
    List<ImageInfo> mImageInfoList;
    BaseFragment mContext;
    RecyclerView mRecyclerView;
    DataApi mDataApi;
    public final static int HEADER = 0;

    public final static int ITEM = 1;

    private List<ImageInfo> adImageInfoList;

    @Override
    public void onImageSaved(ImageInfo imageInfo) {
        reload();

    }

    public void remove(ImageInfo imageInfo) {

        if (mImageInfoList != null) {
            int index = mImageInfoList.indexOf(imageInfo);
            if (index != -1) {
                mImageInfoList.remove(index);
                notifyItemRemoved(index);
            }
        }
        if (mSelectedImageInfoList != null) {
            int index = mSelectedImageInfoList.indexOf(imageInfo);

            if (index != -1) {
                mSelectedImageInfoList.remove(index);
                notifyItemRemoved(index);
            }
        }


    }

    public void setItems(List<ImageInfo> imageInfoList) {

        setAdImages(imageInfoList, layoutManager);
        this.mImageInfoList = imageInfoList;
        notifyDataSetChanged();

    }

    public interface OnUiUpdateListener {
        void onDataChanged(int newAdaptorSize);

        void onImageDeleted(ImageInfo imageInfo);

        void onAllImagesDeleted(List<ImageInfo> mImageInfoList);

    }

    public interface OnULoadingCancelListener {
        void onLoadingCancel();

    }

    OnDetailedScreenLaunchedListener mOnDetailedScreenLaunchedListener;

    public interface OnDetailedScreenLaunchedListener {
        void onDetailScreenLaunched();
    }

    public void setOnDetailedScreenLaunchedListener(OnDetailedScreenLaunchedListener mOnDetailedScreenLaunchedListener) {
        this.mOnDetailedScreenLaunchedListener = mOnDetailedScreenLaunchedListener;
    }

    OnUiUpdateListener mOnUiUpdateListener;
    OnULoadingCancelListener mOnULoadingCancelListener;

    public void setOnUiUpdateListener(OnUiUpdateListener mOnUiUpdateListener) {
        this.mOnUiUpdateListener = mOnUiUpdateListener;
    }

    private List<ImageInfo> getNonAdList(List<ImageInfo> imageInfoList) {
        List<ImageInfo> nonAdImageInfoList = new ArrayList<>();
        for (ImageInfo imageInfo : imageInfoList) {
            if (imageInfo.isAd() == false) {
                nonAdImageInfoList.add(imageInfo);
            }
        }
        return nonAdImageInfoList;
    }

    public void selectAllItems() {
        mSelectedImageInfoList.clear();
        mSelectedImageInfoList.addAll(getNonAdList(mImageInfoList));

//        mSelectedImageInfoList.addAll(mImageInfoList);
        notifyDataSetChanged();

    }

    public void clearAllItems() {
        mSelectedImageInfoList.clear();
        notifyDataSetChanged();

    }

    public void deleteAllItems() {
        mSelectedImageInfoList.clear();
        mImageInfoList.clear();
        notifyDataSetChanged();

    }

    List<ImageInfoLoadingTask> mImageInfoLoadingTasks;

    public void cancelAllTasks() {
        if (mImageInfoLoadingTasks != null) {
            for (int i = 0; i < mImageInfoLoadingTasks.size(); i++) {
                ImageInfoLoadingTask imageInfoLoadingTask = mImageInfoLoadingTasks.get(i);
                if (imageInfoLoadingTask != null) {
                    imageInfoLoadingTask.cancel(true);
                }
            }
        }
    }

    int mCount;
    boolean shownError;

    public void deleteSelectedImages() {
        mCount = 0;
        //cancel loading
        cancelAllTasks();

        final MultipleImageProcessingDialog multipleImageProcessingDialog = new MultipleImageProcessingDialog(
                (imageInfo, pos) -> cancelAllTasks(), mContext.getContext(), null);

        if (mOnULoadingCancelListener != null) {
            mOnULoadingCancelListener.onLoadingCancel();
        }

        mImageInfoLoadingTasks = new ArrayList<>();
        if (mSelectedImageInfoList.size() > 0) {
            multipleImageProcessingDialog.onProcessingStarted(mSelectedImageInfoList.get(0), 1,
                    mSelectedImageInfoList.size());
        }
        final int size = mSelectedImageInfoList.size();

        shownError = false;
        List<ImageInfo> imageInfoList = new ArrayList<>();

        for (ImageInfo imageInfo : mSelectedImageInfoList) {
            if (imageInfo.isAd()) {
                continue;
            }
            String thumb = imageInfo.getAbsoluteThumbFilePath();
            if (imageInfo.getAbsoluteFilePathUri() == null && imageInfo.getImageUri() != null) {
                imageInfo = Utils.getImageInfo(imageInfo, mContext.getContext(), imageInfo.getImageUri(), mDataApi);
            }
            if (imageInfo != null) {
                if (imageInfo.getAbsoluteFilePathUri() != null) {
                    Uri orgUri = imageInfo.getImageUri();
                    imageInfo.setImageUri(imageInfo.getAbsoluteFilePathUri());
                    imageInfoList.add(imageInfo);
                }
            }
        }
        mDataApi.deleteImageFiles((Activity) mContext.getContext(), imageInfoList);
        if (mOnUiUpdateListener != null) {
            List<ImageInfo> list = new ArrayList<>();
            for (ImageInfo imageInfo : imageInfoList) {
                if (!imageInfo.isDeleted()) {
                    list.add(imageInfo);
                }
            }
            mOnUiUpdateListener.onAllImagesDeleted(list);
        }
        multipleImageProcessingDialog.onProcessingDone(mImageInfoList);
    }

    private void showError(int errorId) {

        try {

            String s = mContext.getActivity().getString(errorId);

            Toast.makeText(mContext.getContext(), s, Toast.LENGTH_SHORT).show();
        } catch (Exception e) {

        }
    }

    public int setImageInfo(ImageInfo imageInfo) {

        int index = mImageInfoList.indexOf(imageInfo);
        if (index != -1) {


            mImageInfoList.set(index, imageInfo);
            notifyItemChanged(index);
        }
        return index;
    }


    public interface OnImagesSelectedListener {
        void onAnyImageSelected();

        void onNoneImageSelected();
    }

    OnImagesSelectedListener mOnImagesSelectedListener;
    MyImagesPresenter myImagesPresenter;

    private int appereance;

    private int repeatAfter;
    private GridLayoutManager layoutManager;

    public MyImagesAdaptor(MyImagesPresenter myImagesPresenter, BaseFragment context, List<ImageInfo> imageInfoList,

                           RecyclerView recyclerView, GridLayoutManager layoutManager,
                           OnUiUpdateListener onUiUpdateListener, OnULoadingCancelListener onULoadingCancelListener) {
        this.mOnUiUpdateListener = onUiUpdateListener;
        mContext = context;
        mRecyclerView = recyclerView;
        mRecyclerView.setLayoutManager(layoutManager);
        this.mImageInfoList = imageInfoList;
        this.myImagesPresenter = myImagesPresenter;
        appereance = SettingsManager.getInstance().getGridAppearnece();

        mOnImagesSelectedListener = (OnImagesSelectedListener) context;
        this.mOnULoadingCancelListener = onULoadingCancelListener;
        mDataApi = new FileApi(context.getContext());
        adImageInfoList = new ArrayList<>();
        this.layoutManager = layoutManager;

        int spacingInPixels = context.getResources().getDimensionPixelSize(R.dimen.spacing);
        mRecyclerView.addItemDecoration(new SpacesItemDecoration(spacingInPixels));


        layoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
            @Override
            public int getSpanSize(int position) {
                switch (getItemViewType(position)) {
                    case HEADER:
                        return layoutManager.getSpanCount();

                    case ITEM:
                        return 1;

                    default:
                        return 1;
                }
            }
        });


    }

    private void setAdImages(List<ImageInfo> imageInfoList, GridLayoutManager layoutManager) {


        if (appereance == 0) {//large
            repeatAfter = layoutManager.getSpanCount() * 2;//after these rows show ad

        } else if (appereance == 2) { //small
            repeatAfter = layoutManager.getSpanCount() * 5;

        } else if (appereance == 1) { //normal
            repeatAfter = layoutManager.getSpanCount() * 3;

        }
        repeatAfter = repeatAfter * 2;
        //remove old ads if any

        int size = imageInfoList.size();
        List<ImageInfo> nonAdImageInfoList = new ArrayList<>();
        for (int i = 0; i < imageInfoList.size(); i++) {
            if (imageInfoList.get(i).isAd() == false) {
                nonAdImageInfoList.add(imageInfoList.get(i));
            }
        }
        imageInfoList.clear();
        imageInfoList.addAll(nonAdImageInfoList);

        for (int i = repeatAfter; i < size; i += repeatAfter) {
            ImageInfo imageInfo = new ImageInfo();
            imageInfo.setAd(true);
            imageInfoList.add(i, imageInfo);
            ;

        }
        notifyDataSetChanged();
    }

    public void setLayoutManager(RecyclerView.LayoutManager layoutManager) {
        if (mRecyclerView != null) {
            mRecyclerView.setLayoutManager(layoutManager);
            mRecyclerView.getAdapter().notifyDataSetChanged();
        }

    }


    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {


        if (viewType == ITEM) {

            View view = LayoutInflater.from(mContext.getContext()).inflate(R.layout.mymultiple_image_row, null);

            return new MultipleImagesHolder(view);
        }

        View view = LayoutInflater.from(mContext.getContext()).inflate(R.layout.header_gridview, null);
        HeaderView headerView = new HeaderView(view);

        return headerView;


    }

    static class HeaderView extends RecyclerView.ViewHolder {
        public View rootView;

        public HeaderView(View itemView) {
            super(itemView);
            this.rootView = itemView;

        }
    }

    List<ImageInfo> mSelectedImageInfoList = new ArrayList<>();

    public void setAppereance(int val) {
        appereance = val;
        if (mRecyclerView != null) {
            mRecyclerView.getAdapter().notifyDataSetChanged();
        }
    }


    @Override
    public void onBindViewHolder(final RecyclerView.ViewHolder holder1, int position) {

        ImageInfo imageInfo = mImageInfoList.get(position);


        if (holder1 instanceof MultipleImagesHolder) {
            MultipleImagesHolder holder = (MultipleImagesHolder) holder1;


            RelativeLayout.LayoutParams reLayoutParams = new RelativeLayout.LayoutParams(holder.imageView.getLayoutParams().width, ViewGroup.LayoutParams.MATCH_PARENT);
            boolean hideCompressedInfo = false;

            holder.imageView.setTag(imageInfo);
            int span = layoutManager.getSpanCount();
            if (appereance == 0) {//large

                if (span == 1 || span == 2) {

                    reLayoutParams.height = (int) mContext.getResources().getDimension(R.dimen.large_app);
                } else {
                    reLayoutParams.height = (int) mContext.getResources().getDimension(R.dimen.large_app);
                    reLayoutParams.height = reLayoutParams.height - reLayoutParams.height / 3;
                    hideCompressedInfo = true;
                }
                holder.imageView.setScaleType(ImageView.ScaleType.CENTER_CROP);
                holder.imageView.setLayoutParams(reLayoutParams);
            } else if (appereance == 2) { //small
                hideCompressedInfo = true;
                reLayoutParams.height = (int) mContext.getResources().getDimension(R.dimen.small_app);
                holder.imageView.setLayoutParams(reLayoutParams);
                holder.imageView.setScaleType(ImageView.ScaleType.CENTER_CROP);

            } else if (appereance == 1) { //normal
                if (span == 1 || span == 2) {

                    reLayoutParams.height = (int) mContext.getResources().getDimension(R.dimen.normal_app);
                } else {
                    reLayoutParams.height = (int) mContext.getResources().getDimension(R.dimen.normal_app) / 2;
                    hideCompressedInfo = true;
                }
                holder.imageView.setScaleType(ImageView.ScaleType.CENTER_CROP);
                holder.imageView.setLayoutParams(reLayoutParams);
            }


            if (imageInfo.getAbsoluteThumbFilePath() != null) {
                File file = new File(imageInfo.getAbsoluteThumbFilePath());

                holder.imageView.setImageURI(Uri.fromFile(file));
            } else {
//            holder.webviewOrg.setImageURI(imageInfo.getImageUri());
            }
            if (imageInfo.getWidth() > 0 && imageInfo.getHeight() > 0) {
                holder.resTextView.setText(imageInfo.getWidth() + " x " + imageInfo.getHeight());
            }

            if (imageInfo.getFormatedFileSize() != null && imageInfo.getFormatedFileSize().length() > 0) {

                holder.sizeTextView.setText(String.format("(%s)", imageInfo.getFormatedFileSize()));
            }

            holder.checkBox.setTag(mImageInfoList.get(position));
            holder.checkBox.setOnCheckedChangeListener(null);

            if (mSelectedImageInfoList.contains(mImageInfoList.get(position))) {
                holder.checkBox.setChecked(true);
            } else {
                holder.checkBox.setChecked(false);
            }

            holder.checkBox.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
                @Override
                public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {


                    ImageInfo selImageInfo = (ImageInfo) buttonView.getTag();
                    if (isChecked) {
                        if (mSelectedImageInfoList.contains(selImageInfo) == false) {
                            mSelectedImageInfoList.add(selImageInfo);
                        }
                    } else {
                        mSelectedImageInfoList.remove(selImageInfo);
                    }
                    if (mSelectedImageInfoList.size() > 0) {
                        mOnImagesSelectedListener.onAnyImageSelected();
                    } else {
                        mOnImagesSelectedListener.onNoneImageSelected();
                    }

                }
            });

            holder.imageView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {

                    ImageInfo imageInfoIn = (ImageInfo) holder.imageView.getTag();
                    Intent intent = new Intent();
                    Uri uri = imageInfoIn.getImageUri();
//                        if(uri.toString().startsWith("file:/")) {
//                            uri =mDataApi.getImageUriFromCacheWithFileProvider(imageInfoIn.getDataFile().getName());
//                        }

                    intent.setData(uri);

                    if (mOnDetailedScreenLaunchedListener != null) {
                        mOnDetailedScreenLaunchedListener.onDetailScreenLaunched();
                    }


                    final ImageDetailFragment imageDetailFragment = ImageDetailFragment.newInstance(uri, null);
//                final ResizeFragment resizeFragment = ResizeFragment.newInstance(intent);

//                resizeFragment.setOnImagedSavedListener(MyImagesAdaptor.this);


                    Utils.addFragment((AppCompatActivity) mContext.getActivity(), imageDetailFragment, R.id.contentFrame, true);

                    //   Utils.getImageInfo(mContext.getContext(),Uri.parse("file:///data/user/0/apps.sai.com.imageresizer.demo.debug/files/nonShared/IMG-20180120-WA0005.jpg"),mDataApi);
                }
            });
        } else if (holder1 instanceof HeaderView) {
            HeaderView headerView = (HeaderView) holder1;
//           ViewGroup parentAdView =  headerView.rootView.findViewById(R.id.banner_container_top);
//            if(parentAdView.getChildCount()==0) {
            try {
                String nextid = bannerIds[(count++) % (bannerIds.length - 1)];
                Utils.showFacebookBanner(mContext.getContext(), headerView.rootView, R.id.banner_container_top, nextid);

            } catch (Exception e) {
                e.printStackTrace();
                Utils.showFacebookBanner(mContext.getContext(), headerView.rootView, R.id.banner_container_top, bannerIds[0]);

            }
//            }


        }


    }

    int count = 0;

    String bannerIds[] = new String[]{"179547122769778_189237791800711", "179547122769778_179622146095609", "179547122769778_189046365153187", "179547122769778_189237791800711"};


    @Override
    public int getItemViewType(int position) {
        if (mImageInfoList.get(position).isAd() == false) {
            return ITEM;
        }
        return HEADER;
    }

    private void reload() {
        if (myImagesPresenter != null && mContext != null) {
            try {
                this.mImageInfoList = myImagesPresenter.getImages(mContext.getContext());
                this.mSelectedImageInfoList = new ArrayList<>();
                notifyDataSetChanged();

            } catch (Exception e) {
                e.printStackTrace();
            }

        }

    }

    public List<ImageInfo> getSelectedImageInfoList() {
        return mSelectedImageInfoList;
    }


    @Override
    public int getItemCount() {
        return mImageInfoList.size();
    }

    public static class MultipleImagesHolder extends RecyclerView.ViewHolder {

        View rootView;
        public ImageView imageView;
        public TextView resTextView;
        public TextView sizeTextView;
        public CheckBox checkBox;


        public MultipleImagesHolder(View itemView) {
            super(itemView);
            imageView = itemView.findViewById(R.id.image_multiple);
            resTextView = itemView.findViewById(R.id.text_name_resolution);
            sizeTextView = itemView.findViewById(R.id.text_name_size);
            checkBox = itemView.findViewById(R.id.checkbox);
        }
    }
}

File: app/src/main/java/apps/sai/com/imageresizer/myimages/MyImagesAdaptor.kt
package apps.sai.com.imageresizer.myimages

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.GridLayoutManager.SpanSizeLookup
import androidx.recyclerview.widget.RecyclerView
import apps.sai.com.imageresizer.BaseFragment
import apps.sai.com.imageresizer.R
import apps.sai.com.imageresizer.data.DataApi
import apps.sai.com.imageresizer.data.FileApi
import apps.sai.com.imageresizer.data.ImageInfo
import apps.sai.com.imageresizer.resize.ResizeFragment.OnImagedSavedListener
import apps.sai.com.imageresizer.settings.SettingsManager
import apps.sai.com.imageresizer.util.ImageDetailFragment
import apps.sai.com.imageresizer.util.SpacesItemDecoration
import apps.sai.com.imageresizer.util.Utils
import java.io.File

/**
 * Created by sailesh on 20/01/18.
 */
class MyImagesAdaptor(
    myImagesPresenter: MyImagesPresenter?,
    context: BaseFragment,
    imageInfoList: List<ImageInfo>,
    recyclerView: RecyclerView,
    layoutManager: GridLayoutManager,
    var mOnUiUpdateListener: OnUiUpdateListener?,
    onULoadingCancelListener: OnULoadingCancelListener
) : RecyclerView.Adapter<RecyclerView.ViewHolder>(), OnImagedSavedListener {
    private var mImageInfoList: MutableList<ImageInfo> = mutableListOf()
    private var mContext: BaseFragment = context
    private var mRecyclerView: RecyclerView = recyclerView
    var mDataApi: DataApi
    private val adImageInfoList: List<ImageInfo>

    override fun onImageSaved(imageInfo: ImageInfo) {
        reload()
    }

    fun remove(imageInfo: ImageInfo) {
        val index1 = mImageInfoList.indexOf(imageInfo)
        if (index1 != -1) {
            mImageInfoList.removeAt(index1)
            notifyItemRemoved(index1)
        }
        val index = mSelectedImageInfoList.indexOf(imageInfo)

        if (index != -1) {
            mSelectedImageInfoList.removeAt(index)
            notifyItemRemoved(index)
        }
    }

    fun setItems(imageInfoList: MutableList<ImageInfo>) {
        setAdImages(imageInfoList, layoutManager)
        this.mImageInfoList = imageInfoList
        notifyDataSetChanged()
    }

    interface OnUiUpdateListener {
        fun onDataChanged(newAdaptorSize: Int)
        fun onImageDeleted(imageInfo: ImageInfo)
        fun onAllImagesDeleted(mImageInfoList: List<ImageInfo>)
    }

    interface OnULoadingCancelListener {
        fun onLoadingCancel()
    }

    private var mOnDetailedScreenLaunchedListener: OnDetailedScreenLaunchedListener? = null

    interface OnDetailedScreenLaunchedListener {
        fun onDetailScreenLaunched()
    }

    fun setOnDetailedScreenLaunchedListener(mOnDetailedScreenLaunchedListener: OnDetailedScreenLaunchedListener?) {
        this.mOnDetailedScreenLaunchedListener = mOnDetailedScreenLaunchedListener
    }

    private var mOnULoadingCancelListener: OnULoadingCancelListener?

    fun setOnUiUpdateListener(mOnUiUpdateListener: OnUiUpdateListener?) {
        this.mOnUiUpdateListener = mOnUiUpdateListener
    }

    private fun getNonAdList(imageInfoList: List<ImageInfo>?): List<ImageInfo> {
        val nonAdImageInfoList: MutableList<ImageInfo> = ArrayList()
        for (imageInfo in imageInfoList!!) {
            if (imageInfo.isAd == false) {
                nonAdImageInfoList.add(imageInfo)
            }
        }
        return nonAdImageInfoList
    }

    fun selectAllItems() {
        mSelectedImageInfoList.clear()
        mSelectedImageInfoList.addAll(getNonAdList(mImageInfoList))
        notifyDataSetChanged()
    }

    fun clearAllItems() {
        mSelectedImageInfoList.clear()
        notifyDataSetChanged()
    }

    fun deleteAllItems() {
        mSelectedImageInfoList.clear()
        mImageInfoList.clear()
        notifyDataSetChanged()
    }

    private var mCount: Int = 0
    private var shownError: Boolean = false

    fun deleteSelectedImages() {
        mCount = 0
        if (mOnULoadingCancelListener != null) {
            mOnULoadingCancelListener!!.onLoadingCancel()
        }
        shownError = false
        val imageInfoList: MutableList<ImageInfo> = ArrayList()

        for (imageInfo in mSelectedImageInfoList) {
            if (imageInfo.isAd) {
                continue
            }
            if (imageInfo.imageUri != null) {
                val imageInfoProcessed =
                    Utils.getImageInfo(imageInfo, mContext.context, imageInfo.imageUri, mDataApi)
                imageInfoProcessed.imageUri = imageInfo.absoluteFilePathUri
                imageInfoList.add(imageInfoProcessed)
            }
        }
        mDataApi.deleteImageFiles(mContext.context as Activity?, imageInfoList)
        if (mOnUiUpdateListener != null) {
            val list: MutableList<ImageInfo> = ArrayList()
            for (imageInfo in imageInfoList) {
                if (!imageInfo.isDeleted) {
                    list.add(imageInfo)
                }
            }
            mOnUiUpdateListener!!.onAllImagesDeleted(list)
        }
    }

    private fun showError(errorId: Int) {
        try {
            val s = mContext.requireActivity().getString(errorId)
            Toast.makeText(mContext.context, s, Toast.LENGTH_SHORT).show()
        } catch (e: Exception) {
        }
    }

    fun setImageInfo(imageInfo: ImageInfo): Int {
        val index = mImageInfoList.indexOf(imageInfo)
        if (index != -1) {
            mImageInfoList[index] = imageInfo
            notifyItemChanged(index)
        }
        return index
    }

    interface OnImagesSelectedListener {
        fun onAnyImageSelected()
        fun onNoneImageSelected()
    }

    private var mOnImagesSelectedListener: OnImagesSelectedListener
    private var myImagesPresenter: MyImagesPresenter?

    private var appereance: Int

    private var repeatAfter = 0
    private val layoutManager: GridLayoutManager

    private fun setAdImages(
        imageInfoList: MutableList<ImageInfo>,
        layoutManager: GridLayoutManager
    ) {
        if (appereance == 0) { //large
            repeatAfter = layoutManager.spanCount * 2 //after these rows show ad
        } else if (appereance == 2) { //small
            repeatAfter = layoutManager.spanCount * 5
        } else if (appereance == 1) { //normal
            repeatAfter = layoutManager.spanCount * 3
        }
        repeatAfter *= 2

        //remove old ads if any
        val size = imageInfoList.size
        val nonAdImageInfoList: MutableList<ImageInfo> = ArrayList()
        for (i in imageInfoList.indices) {
            if (!imageInfoList[i].isAd) {
                nonAdImageInfoList.add(imageInfoList[i])
            }
        }
        imageInfoList.clear()
        imageInfoList.addAll(nonAdImageInfoList)
        var i = repeatAfter
        while (i < size) {
            val imageInfo = ImageInfo()
            imageInfo.isAd = true
            imageInfoList.add(i, imageInfo)
            i += repeatAfter
        }
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        if (viewType == ITEM) {
            val view =
                LayoutInflater.from(mContext.context).inflate(R.layout.mymultiple_image_row, null)

            return MultipleImagesHolder(view)
        }

        val view = LayoutInflater.from(mContext.context).inflate(R.layout.header_gridview, null)
        val headerView = HeaderView(view)

        return headerView
    }

    internal class HeaderView(var rootView: View) : RecyclerView.ViewHolder(rootView)

    private var mSelectedImageInfoList: MutableList<ImageInfo> = ArrayList()
    override fun onBindViewHolder(holder1: RecyclerView.ViewHolder, position: Int) {
        val imageInfo = mImageInfoList[position]
        if (holder1 is MultipleImagesHolder) {
            val reLayoutParams = RelativeLayout.LayoutParams(
                holder1.imageView.layoutParams.width,
                ViewGroup.LayoutParams.MATCH_PARENT
            )
            holder1.imageView.tag = imageInfo
            val span = layoutManager.spanCount
            if (appereance == 0) { //large

                if (span == 1 || span == 2) {
                    reLayoutParams.height =
                        mContext.resources.getDimension(R.dimen.large_app).toInt()
                } else {
                    reLayoutParams.height =
                        mContext.resources.getDimension(R.dimen.large_app).toInt()
                    reLayoutParams.height = reLayoutParams.height - reLayoutParams.height / 3
                }
                holder1.imageView.scaleType = ImageView.ScaleType.CENTER_CROP
                holder1.imageView.layoutParams = reLayoutParams
            } else if (appereance == 2) { //small
                reLayoutParams.height = mContext.resources.getDimension(R.dimen.small_app).toInt()
                holder1.imageView.layoutParams = reLayoutParams
                holder1.imageView.scaleType = ImageView.ScaleType.CENTER_CROP
            } else if (appereance == 1) { //normal
                if (span == 1 || span == 2) {
                    reLayoutParams.height =
                        mContext.resources.getDimension(R.dimen.normal_app).toInt()
                } else {
                    reLayoutParams.height =
                        mContext.resources.getDimension(R.dimen.normal_app).toInt() / 2
                }
                holder1.imageView.scaleType = ImageView.ScaleType.CENTER_CROP
                holder1.imageView.layoutParams = reLayoutParams
            }
            if (imageInfo.absoluteThumbFilePath != null) {
                val file = File(imageInfo.absoluteThumbFilePath)
                holder1.imageView.setImageURI(Uri.fromFile(file))
            }
            if (imageInfo.width > 0 && imageInfo.height > 0) {
                holder1.resTextView.text = imageInfo.width.toString() + " x " + imageInfo.height
            }

            if (imageInfo.formatedFileSize != null && imageInfo.formatedFileSize.length > 0) {
                holder1.sizeTextView.text = String.format("(%s)", imageInfo.formatedFileSize)
            }

            holder1.checkBox.tag = mImageInfoList[position]
            holder1.checkBox.setOnCheckedChangeListener(null)

            holder1.checkBox.isChecked = mSelectedImageInfoList.contains(mImageInfoList[position])

            holder1.checkBox.setOnCheckedChangeListener { buttonView, isChecked ->
                val selImageInfo = buttonView.tag as ImageInfo
                if (isChecked) {
                    if (!mSelectedImageInfoList.contains(selImageInfo)) {
                        mSelectedImageInfoList.add(selImageInfo)
                    }
                } else {
                    mSelectedImageInfoList.remove(selImageInfo)
                }
                if (mSelectedImageInfoList.size > 0) {
                    mOnImagesSelectedListener.onAnyImageSelected()
                } else {
                    mOnImagesSelectedListener.onNoneImageSelected()
                }
            }

            holder1.imageView.setOnClickListener {
                val imageInfoIn = holder1.imageView.tag as ImageInfo
                val intent = Intent()
                val uri = imageInfoIn.imageUri
                intent.setData(uri)

                if (mOnDetailedScreenLaunchedListener != null) {
                    mOnDetailedScreenLaunchedListener!!.onDetailScreenLaunched()
                }


                val imageDetailFragment = ImageDetailFragment.newInstance(uri, null)

                Utils.addFragment(
                    mContext.activity as AppCompatActivity?,
                    imageDetailFragment,
                    R.id.contentFrame,
                    true
                )
            }
        } else if (holder1 is HeaderView) {
            try {
                val next = bannerIds[count++ % (bannerIds.size - 1)]
                Utils.showFacebookBanner(
                    mContext.context,
                    holder1.rootView,
                    R.id.banner_container_top,
                    next
                )
            } catch (e: Exception) {
                e.printStackTrace()
                Utils.showFacebookBanner(
                    mContext.context,
                    holder1.rootView,
                    R.id.banner_container_top,
                    bannerIds[0]
                )
            }
        }
    }

    private var count: Int = 0

    private var bannerIds: Array<String> = arrayOf(
        "179547122769778_189237791800711",
        "179547122769778_179622146095609",
        "179547122769778_189046365153187",
        "179547122769778_189237791800711"
    )


    init {
        mRecyclerView.layoutManager = layoutManager
        this.mImageInfoList.addAll(imageInfoList)
        this.myImagesPresenter = myImagesPresenter
        appereance = SettingsManager.getInstance().gridAppearnece
        mOnImagesSelectedListener = context as OnImagesSelectedListener
        this.mOnULoadingCancelListener = onULoadingCancelListener
        mDataApi = FileApi(context.context)
        adImageInfoList = ArrayList()
        this.layoutManager = layoutManager
        val spacingInPixels = context.resources.getDimensionPixelSize(R.dimen.spacing)
        mRecyclerView.addItemDecoration(SpacesItemDecoration(spacingInPixels))
        layoutManager.spanSizeLookup = object : SpanSizeLookup() {
            override fun getSpanSize(position: Int): Int {
                return when (getItemViewType(position)) {
                    HEADER -> layoutManager.spanCount
                    ITEM -> 1
                    else -> 1
                }
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        if (!mImageInfoList[position].isAd) {
            return ITEM
        }
        return HEADER
    }


    private fun reload() {
        if (myImagesPresenter != null) {
            try {
                // this.mImageInfoList = myImagesPresenter.getImages(mContext.getContext());
                this.mSelectedImageInfoList = ArrayList()
                notifyDataSetChanged()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    val selectedImageInfoList: List<ImageInfo>
        get() = mSelectedImageInfoList


    override fun getItemCount(): Int {
        return mImageInfoList.size
    }

    class MultipleImagesHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var imageView: ImageView = itemView.findViewById(R.id.image_multiple)
        var resTextView: TextView = itemView.findViewById(R.id.text_name_resolution)
        var sizeTextView: TextView = itemView.findViewById(R.id.text_name_size)
        var checkBox: CheckBox = itemView.findViewById(R.id.checkbox)
    }

    companion object {
        const val HEADER: Int = 0
        const val ITEM: Int = 1
    }
}
